api/processors/services/group_builder.py

```diff
@@ -146,23 +150,120 @@ def _get_allowed_group_size(age_range: AgeRange) -> GroupSizeRestriction:
         raise ValueError(f"Group age range is inconsistent with boundaries: {age_range}")
 
     @staticmethod
-    def _get_group_candidate(teacher_id: int) -> GroupCandidate | None:
-        # THIS IS A STUB METHOD RETURNING DUMMY VALUES
-        # TODO filter for language+level, communication language, age groups, time slots (all UTC?)
-        # TODO iterate in correct priority order, yield results
+    def _get_best_group_candidate(teacher_id: int) -> GroupCandidate | None:
         teacher = Teacher.objects.get(pk=teacher_id)
+        group_candidates: list[GroupCandidate] = []
+
+        for language_and_level in teacher.teaching_languages_and_levels.all():
+            logger.info(language_and_level)
+            for first_time_slot, second_time_slot in GroupBuilder._iterate_lesson_times(teacher):
+                for age_range in teacher.student_age_ranges.all():
+                    group_candidate = GroupBuilder._get_group_candidate(
+                        teacher=teacher,
+                        age_range=age_range,
+                        language_and_level=language_and_level,
+                        day_time_slots=(first_time_slot, second_time_slot),
+                    )
+                    if group_candidate is not None:
+                        group_candidates.append(group_candidate)
+
+        if len(group_candidates) == 0:
+            return None
+
+        group_candidates.sort(key=lambda candidate: candidate.priority, reverse=True)
+        return group_candidates[0]
+
+    @staticmethod
+    def _get_group_candidate(
+        teacher: Teacher,
+        age_range: AgeRange,
+        language_and_level: LanguageAndLevel,
+        day_time_slots: Collection[DayAndTimeSlot],
+    ) -> GroupCandidate | None:
+        restrictions = GroupBuilder._get_allowed_group_size(age_range)
+        communication_language = CommunicationLanguageMode(
+            teacher.personal_info.communication_language_mode
+        )
+        students = GroupBuilder._get_students(
+            language_and_level=language_and_level,
+            age_range=age_range,
+            communication_language=communication_language,
+            day_time_slots=day_time_slots,
+            max_students_num=restrictions.max,
+        )
+        logger.debug(students)
+        if len(students) < restrictions.min:
+            return None
```

Comment:
```suggestion
        if len(students) < restrictions.min:
            logger.debug("Not enough students")
            return None
```
I know it's strange to insert debug commands, but I couldn't help it :)

Reply:
(nit)

Reply:
Errrm, that's just a lot of logs. Imagine we try every possible combination of language, level, date time and most of them don't work (maybe not on real data in fact! But in our testing database it's a bit of pain to find someting useful :cry: )